src/toxcore/dht_new/packet.rs

```diff
@@ -0,0 +1,938 @@
+    GetNodes(GetNodes),
+    /// [`SendNodes`](./struct.SendNodes.html) structure.
+    SendNodes(SendNodes),
+}
+
+impl ToBytes for DhtPacket {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        match *self {
+            DhtPacket::PingReq(ref p) => p.to_bytes(buf),
+            DhtPacket::PingResp(ref p) => p.to_bytes(buf),
+            DhtPacket::GetNodes(ref p) => p.to_bytes(buf),
+            DhtPacket::SendNodes(ref p) => p.to_bytes(buf),
+        }
+    }
+}
+
+impl FromBytes for DhtPacket {
+    named!(from_bytes<DhtPacket>, alt!(
+        map!(PingReq::from_bytes, DhtPacket::PingReq) |
+        map!(PingResp::from_bytes, DhtPacket::PingResp) |
+        map!(GetNodes::from_bytes, DhtPacket::GetNodes) |
+        map!(SendNodes::from_bytes, DhtPacket::SendNodes)
+    ));
+}
+
+/** DHT Request packet.
+
+https://zetok.github.io/tox-spec/#dht-request-packets
+*/
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum DhtRequest {
+    /// [`NatPingReq`](./struct.NatPingReq.html) structure.
+    NatPingReq(PingReq),
+    /// [`NatPingResp`](./struct.NatPingResp.html) structure.
+    NatPingResp(PingResp),
+}
+
+/** NatPing request of DHT Request packet.
+*/
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct NatPingReq(pub PingReq);
+
+/** NatPing response of DHT Request packet.
+*/
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct NatPingResp(pub PingResp);
+
+impl ToBytes for DhtRequest {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        match *self {
+            DhtRequest::NatPingReq(ref p) => p.to_bytes(buf),
+            DhtRequest::NatPingResp(ref p) => p.to_bytes(buf),
+        }
+    }
+}
+
+impl FromBytes for DhtRequest {
+    named!(from_bytes<DhtRequest>, alt!(
+        map!(PingReq::from_bytes, DhtRequest::NatPingReq) |
+        map!(PingResp::from_bytes, DhtRequest::NatPingResp)
+    ));
+}
+
+impl FromBytes for NatPingReq {
+    named!(from_bytes<NatPingReq>, do_parse!(
+        tag!(&[0xfe][..]) >>
+        ping: call!(PingReq::from_bytes) >>
+        (NatPingReq(ping))
+    ));
+}
+
+impl ToBytes for NatPingReq {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        let &NatPingReq(x) = self;
+        do_gen!(buf,
+            gen_be_u8!(0xfe) >>
+            gen_call!(|buf, data| PingReq::to_bytes(data, buf), &x)
+        )
+    }
+}
+
+impl FromBytes for NatPingResp {
+    named!(from_bytes<NatPingResp>, do_parse!(
+        tag!(&[0xfe][..]) >>
+        ping: call!(PingResp::from_bytes) >>
+        (NatPingResp(ping))
+    ));
+}
+
+impl ToBytes for NatPingResp {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        let &NatPingResp(x) = self;
+        do_gen!(buf,
+            gen_be_u8!(0xfe) >>
+            gen_call!(|buf, data| PingResp::to_bytes(data, buf), &x)
```

Comment:
`&x` -> `&self.0`. It would be better if you declare PingReq in `pub struct NatPingReq(pub PingReq);` not as an unnamed struct field, but as named `ping`.